File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    length = len(numbers)
    answer = [-1] * len(numbers)
    stack = []
    
    for i in range(length - 1, -1, -1):
        while stack and stack[-1] <= numbers[i]:
            stack.pop()
        if stack:
            answer[i] = stack[-1]
        stack.append(numbers[i])
    
    return answer


# 시간초과
# def solution(numbers):
#     answer = []
    
#     for i in range(len(numbers)):
#         big = False
#         for j in range(i, len(numbers)):
#             if numbers[i] < numbers[j]:
#                 big = True
#                 answer.append(numbers[j])
#                 break
#         if not big:
#             answer.append(-1)
    
#     return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int len = numbers.length;
        Stack<Integer> stack = new Stack<>();
        int[] answer = new int[len];
        for (int i = 0; i < len; i++)
            answer[i] = -1;
        
        int idx = len-2;
        stack.push(numbers[len-1]);
        while (idx >= 0) {
            int number = numbers[idx];
            int stackNum = stack.peek();
            if (number < stackNum) {
                answer[idx] = stackNum;
                idx--;
                stack.push(number);
            }
            else {
                stack.pop();
            }
            if (stack.isEmpty()) {
                stack.push(number);
                idx--;
            }
        }
        
        return answer;
    }
}